Overhaul exercise for 2024 course

adjavon opened the pull request:
See issues #7 #8 #9 and #10.

The goal of this PR is to change the exercise so that the bulk of it works on a slightly modified version of MNIST that is classified by color instead of by digit. 
Other changes include: 
- [x] Adding an overview to the README
- [x] Using networks they have seen already in previous exercises, like the `UNet` and the `DenseModel` (from the failure modes exercise)
- [x] Simplifying the GAN training so that it can be done within the bounds of the exercise
- [x] Providing the scripts used to train the model weights that we provide them with
- [x] Modifying the text to build the story around this simpler problem
- [ ] Adding synapse prediction/other biological problems as a bonus exercise

adjavon commented:
- [ ] Add GitHub Action to build the notebook

afoix commented:
This comment is general feedback I find while checking the exercise. This is not to do with code but rather with the flow of the exercise. 

## Part 1
- [ ] It seems that the image used in the `README.md` shows a Coloured MNIST dataset with classes that contain multiple colours. For example, the first two numbers in the  top left belong to class "2" but are respectively a 5 yellow and a 0 red. How would you use the classes to classify by colors if they are not related? 
<img width="887" alt="Screenshot 2024-08-17 at 11 07 58" src="https://github.com/user-attachments/assets/52bb7748-27be-44b9-ac34-89dc7b63265f">

- [ ] The dataset fails to completely download because of `HTTP Error 403: Forbidden`:
<img width="1026" alt="Screenshot 2024-08-17 at 11 08 25" src="https://github.com/user-attachments/assets/ea6b7bac-3a79-4c24-988b-6fdb77321109">


### Task 1.1: Load the classifier
- [ ] In the loader classifier, we are assuming that they have this path: `checkpoint = torch.load("extras/checkpoints/model.pth")` and this file does not seem commited. (I've commented out the checkpoint loading for the sake of continuing with the review)
<img width="918" alt="Screenshot 2024-08-17 at 11 22 07" src="https://github.com/user-attachments/assets/49835858-ab17-4ede-a58a-9e244e66c55e">

## Part 2

### Task 2.1 Get an attribution

- [ ] In the task 2.1 it looks like the attribution visualisation gives a black images only:
<img width="969" alt="Screenshot 2024-08-17 at 11 26 54" src="https://github.com/user-attachments/assets/d56f4b31-0c17-4c93-980a-90cee544f35f">

Looks like it can be fixed with the following (normalising):
```
    #ax2.imshow(np.abs(attribution))
    ax2.imshow(np.abs(attribution)/np.abs(attribution).max())
```
<img width="755" alt="Screenshot 2024-08-17 at 12 00 41" src="https://github.com/user-attachments/assets/38e06405-1dee-4162-87a6-3b57d1609b4b">

- [ ] Maybe a separate plot for the adverserial loss as it's on a quite different scale as the other two? (or maybe it's fine, I guess this is just taste)
<img width="560" alt="Screenshot 2024-08-17 at 12 13 24" src="https://github.com/user-attachments/assets/16520d82-2939-42cc-b245-4f28802678b9">

## Part4
### Task 4: Create counterfactuals
- [ ] I am not sure if the confusion matrix is expected to look like this:
<img width="593" alt="Screenshot 2024-08-17 at 12 17 54" src="https://github.com/user-attachments/assets/29ffd8ad-d1e4-4a4f-a4b3-27d8a01069a5">


### in function `visualize_color_attribution_and_counterfactual`
- [ ] same visualisation issue as with task 2.1, can be fixed with `ax2.imshow(np.abs(attribution)/np.abs(attribution).max())`

## In the readme
- [ ] I believe the way to run jupyter lab is without having activated the environment in the first place. (We create the env, but we run jupyter lab from base which has it installed, unlike our env, and then this guy finds out our env for us to select in the GUI)
- [ ] an extra dataset is mentioned in the readme as a "if time permits" exercise:
<img width="900" alt="Screenshot 2024-08-17 at 12 24 38" src="https://github.com/user-attachments/assets/cd13aa36-e5f4-4995-8815-a012c338b19d">
It does not appear in the notebook.

afoix reviewed (commented):
Review done and comments added at the end of the PR 😊

adjavon commented:
Thanks Anna for the review! Sorry about the setup issues. I've fixed almost everything that you found in the commits above. Some notes about what fixes what:
- The setup file now uses conda and runs the classification so we can load the weights
- The HTTPError seems to be just one of three possible URLs to download from & I've checked that the full dataset is downlaoded
- The confusion matrix is normal again once the classifier is available
- I've also cleaned up the README and the plotting.

The only thing missing is that I realized that the normalization you suggested is indeed necessary, but it breaks one of the points I'm making 🤯. All for the better though, since the resulting narrative makes more sense (also with the lecture). I'm working on that now and I'll ping you when it's done!!

As for the colors, yes, it's supposed to be more than one color per class! The answer to what makes the classes is hidden at the very end of the solution notebook 😉

adjavon commented on README.md:
```diff
@@ -1,19 +1,39 @@
+We will be working with a simple example which is a fun derivation on the MNIST dataset that you will have seen in previous exercises in this course. 
+Unlike regular MNIST, our dataset is classified not by number, but by color!
+
+![CMNIST](assets/cmnist.png)
```
Check the very end of the solutions notebook!! 😈 🌈 🙊
